### ISSUE TITLE
[IMPLEMENTATION] Replace Stub BitNet Transformer Forward Pass with Production-Ready Implementation

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#360]
# [IMPLEMENTATION] Replace Stub BitNet Transformer Forward Pass with Production-Ready Implementation

## Problem Description

The BitNet.rs project currently has multiple stub implementations and simplified forward pass mechanisms that need to be replaced with production-ready, quantized neural network inference implementations. Analysis of the codebase reveals several critical gaps between the existing transformer implementation and the requirements for a production BitNet inference system.

## Environment

- **Codebase**: BitNet.rs neural network inference framework
- **Rust Version**: 1.90.0+ (2024 edition)
- **Features Required**: `--no-default-features --features cpu` or `gpu`
- **Affected Modules**:
  - `crates/bitnet-inference/src/simple_forward.rs`
  - `crates/bitnet-inference/src/layers/attention.rs`
  - `crates/bitnet-models/src/transformer.rs`
  - Related quantization and kernel modules

### Current Implementation Issues

1. **Simplified Forward Pass**: The `simple_forward.rs` provides only basic embedding + lm_head functionality without attention mechanisms
2. **Simulation vs Production**: Multiple stub implementations exist that provide simplified attention mask application and incomplete quantization paths
3. **Missing Quantized Transformer Integration**: Gap between the full-precision transformer in `transformer.rs` and quantized BitNet requirements
4. **Incomplete Production Validation**: Simulation implementations don't fully validate production inference accuracy

### Technical Gaps Identified

1. **BitNet Attention Integration**: Current attention implementation lacks proper integration with BitNet 1-bit quantization
2. **Quantized Linear Layers**: Transformer blocks need full integration with I2_S, TL1, TL2 quantization methods
3. **CUDA Kernel Integration**: Missing optimized quantized transformer forward pass for GPU acceleration
4. **Cross-Validation Framework**: Need systematic validation against C++ reference implementation

## Impact Assessment

- **Severity**: High - Core inference functionality incomplete
- **Affected Users**: All users attempting real BitNet inference
- **Performance Impact**: Current implementation cannot achieve BitNet's efficiency benefits
- **Business Impact**: Framework cannot deliver on core value proposition of 1-bit quantized inference

## Reproduction Steps

1. **Build with quantization features**:
   ```bash
   cargo build --no-default-features --features cpu
   ```

2. **Attempt quantized inference**:
   ```bash
   cargo run -p xtask -- infer --model path/to/model.gguf --prompt "Test inference"
   ```

3. **Observe issues**:
   - Simple forward pass doesn't use full attention mechanisms
   - Quantized layers not properly integrated into transformer forward pass
   - Missing production-grade attention mask handling
   - Incomplete KV-cache integration with quantized attention

### Primary Implementation Approach

**Phase 1: Core BitNet Transformer Forward Pass**

1. **Create Production BitNetTransformer Module**:
   ```rust
   // crates/bitnet-inference/src/transformer/mod.rs
   pub struct BitNetTransformer {
       config: BitNetConfig,
       embeddings: QuantizedEmbedding,
       layers: Vec<BitNetTransformerBlock>,
       norm: LayerNorm,
       lm_head: QuantizedLinear,
       device: Device,
   }

   impl BitNetTransformer {
       pub async fn forward(
           &self,
           input_ids: &BitNetTensor,
           attention_mask: Option<&BitNetTensor>,
           kv_cache: Option<&mut KVCache>,
       ) -> Result<BitNetTensor>;
   }
   ```

2. **Integrate Quantized Attention Mechanisms**:
   - Replace simulation attention mask with production causal masking
   - Implement proper GQA (Grouped Query Attention) with quantized projections
   - Add optimized rotary embedding application for quantized tensors

3. **Implement BitNetTransformerBlock**:
   ```rust
   pub struct BitNetTransformerBlock {
       attention: BitNetAttention,
       mlp: BitNetMLP,
       input_layernorm: LayerNorm,
       post_attention_layernorm: LayerNorm,
   }
   ```

**Phase 2: Quantization Integration**

1. **Quantized MLP Implementation**:
   - Integrate I2_S quantization for feed-forward layers
   - Implement SwiGLU activation with quantized projections
   - Add proper scaling and dequantization paths

2. **Device-Aware Kernel Selection**:
   - Automatic CPU (SIMD) vs GPU (CUDA) kernel selection
   - Fallback chains for unsupported configurations
   - Performance-optimized quantized matrix multiplication

3. **Memory-Efficient KV-Cache**:
   - Integrate quantized KV-cache with transformer blocks
   - Implement dynamic cache growth for variable sequence lengths
   - Add cache compression for long sequences

**Phase 3: Production Validation**

1. **Cross-Validation Framework Integration**:
   ```bash
   export BITNET_GGUF="path/to/model.gguf"
   cargo run -p xtask -- crossval --implementation bitnet-transformer
   ```

2. **Accuracy Validation**:
   - Systematic comparison with C++ reference implementation
   - Validate quantization accuracy meets 99%+ threshold
   - Performance benchmarking against baseline implementations

### Alternative Approaches Considered

1. **Gradual Migration**: Replace simple_forward.rs incrementally
   - **Pros**: Lower risk, easier testing
   - **Cons**: Prolonged incomplete state, complex migration path

2. **Separate Quantized Branch**: Maintain both FP32 and quantized paths
   - **Pros**: Maintains compatibility
   - **Cons**: Code duplication, maintenance overhead

3. **FFI Bridge Enhancement**: Extend C++ integration
   - **Pros**: Leverage existing C++ implementation
   - **Cons**: Rust-native goals not achieved, deployment complexity

### Task Breakdown

**High Priority (Week 1-2)**
- [ ] **AC1**: Create `BitNetTransformer` struct with quantized layer integration
- [ ] **AC2**: Implement `forward()` method with proper attention mechanisms
- [ ] **AC3**: Replace `simple_forward.rs` usage in tests with full transformer
- [ ] **AC4**: Integrate I2_S quantization into transformer blocks

**Medium Priority (Week 3-4)**
- [ ] **AC5**: Implement production-grade attention mask handling
- [ ] **AC6**: Add GQA support with quantized K/V projections
- [ ] **AC7**: Integrate optimized CUDA kernels for GPU inference
- [ ] **AC8**: Add comprehensive error handling and validation

**Low Priority (Week 5-6)**
- [ ] **AC9**: Implement dynamic KV-cache growth
- [ ] **AC10**: Add cross-validation against C++ reference
- [ ] **AC11**: Performance optimization and benchmarking
- [ ] **AC12**: Documentation and integration examples

### Testing Strategy

1. **Unit Tests**:
   ```bash
   cargo test --package bitnet-inference --test transformer_tests --no-default-features --features cpu
   ```

2. **Integration Tests**:
   ```bash
   cargo test --workspace --no-default-features --features cpu -- test_bitnet_transformer_inference
   ```

3. **Cross-Validation**:
   ```bash
   cargo run -p xtask -- crossval --tolerance 1e-4
   ```

4. **Performance Benchmarks**:
   ```bash
   cargo run -p xtask -- bench --model microsoft/bitnet-b1.58-2B-4T-gguf
   ```

### Risk Assessment

**High Risk**:
- Quantization accuracy regression during migration
- Performance degradation compared to current implementation
- CUDA kernel integration complexity

**Medium Risk**:
- KV-cache compatibility with existing inference pipeline
- Memory usage increase during transition period
- Cross-validation framework integration

**Low Risk**:
- API compatibility (can maintain backward compatibility)
- Documentation updates
- Test suite expansion

**Mitigation Strategies**:
- Implement feature flag `--features quantized-transformer` for gradual rollout
- Maintain regression tests against current accuracy baselines
- Use cross-validation framework for continuous validation
- Implement comprehensive benchmarking suite

### Definition of Done

1. **Functional Requirements**:
   - [ ] Full BitNet transformer forward pass with I2_S quantization
   - [ ] Production-grade attention mechanisms with proper masking
   - [ ] Integrated KV-cache for autoregressive generation
   - [ ] Device-aware kernel selection (CPU SIMD / GPU CUDA)

2. **Quality Requirements**:
   - [ ] 99%+ accuracy compared to FP32 reference implementation
   - [ ] Cross-validation passes with C++ BitNet reference
   - [ ] Performance meets or exceeds current simple_forward.rs baseline
   - [ ] Memory usage within 10% of theoretical optimum

3. **Integration Requirements**:
   - [ ] Compatible with existing GGUF model loading pipeline
   - [ ] Works with current tokenizer integration
   - [ ] Supports all quantization methods (I2_S, TL1, TL2)
   - [ ] Maintains API compatibility for inference clients

4. **Documentation Requirements**:
   - [ ] Architecture documentation updated for quantized transformer
   - [ ] Performance benchmarking results documented
   - [ ] Migration guide from simple_forward.rs to full transformer
   - [ ] Cross-validation procedures documented

### Validation Commands

```bash
# Build with quantized transformer
cargo build --no-default-features --features cpu,quantized-transformer

# Run full test suite
cargo test --workspace --no-default-features --features cpu

# Cross-validate accuracy
export BITNET_GGUF="microsoft/bitnet-b1.58-2B-4T-gguf/model.gguf"
cargo run -p xtask -- crossval --implementation bitnet-transformer

# Performance benchmark
cargo run -p xtask -- bench --implementation bitnet-transformer --baseline simple-forward

# Integration test with real model
cargo run -p xtask -- infer --model $BITNET_GGUF --prompt "The BitNet architecture" --max-tokens 100
```

## Related Issues

This issue consolidates multiple related stubs and simulations:

- **Attention Mechanisms**: Issues related to `BitNetAttention::compute_attention` mask application
- **Quantization Integration**: Dead code cleanup for I2_S, TL1, TL2 quantizers
- **CUDA Kernels**: GPU-optimized quantized linear layer implementations
- **KV-Cache**: Production-ready cache management and dynamic growth
- **Cross-Validation**: Systematic validation against C++ reference implementation

### Memory Layout
- Quantized weights: 1-bit + scale factors per channel/group
- KV-cache: FP16 precision for memory efficiency
- Activations: Mixed precision (FP16/BF16 on GPU, FP32 on CPU)

### Performance Targets
- **Inference Speed**: 2x faster than FP32 baseline (theoretical 8x from quantization)
- **Memory Usage**: 8x reduction in model weights, 2x reduction in activations
- **Accuracy**: >99% preserved compared to FP32 reference

### Compatibility Matrix
| Feature Flag | Quantization | Device | Kernels |
|--------------|-------------|---------|---------|
| `cpu` | I2_S, TL1 | CPU | SIMD (AVX2/NEON) |
| `gpu` | I2_S, TL2 | CUDA | Mixed Precision |
| `ffi` | IQ2_S | CPU/GPU | C++ Reference |

---

**Priority**: High
**Complexity**: High
**Estimated Effort**: 4-6 weeks
**Dependencies**: Quantization kernels, GGUF loading, tokenizer integration
